src/mpi/datatype/typeutil.c

```diff
@@ -164,16 +232,17 @@ int MPII_create_pairtypes(void)
         /* we use the _unsafe version because we are still in MPI_Init, before
          * multiple threads are permitted and possibly before support for
          * critical sections is entirely setup */
-        ptr = (MPIR_Datatype *) MPIR_Handle_obj_alloc_unsafe(&MPIR_Datatype_mem);
+        dptr = (MPIR_Datatype *) MPIR_Handle_obj_alloc_unsafe(&MPIR_Datatype_mem);
 
-        MPIR_Assert(ptr);
-        MPIR_Assert(ptr->handle == mpi_pairtypes[i]);
+        MPIR_Assert(dptr);
+        MPIR_Assert(dptr->handle == mpi_pairtypes[i].dtype);
         /* this is a redundant alternative to the previous statement */
-        MPIR_Assert((void *) ptr ==
-                    (void *) (MPIR_Datatype_direct + HANDLE_INDEX(mpi_pairtypes[i])));
+        MPIR_Assert((void *) dptr ==
+                    (void *) (MPIR_Datatype_direct + HANDLE_INDEX(mpi_pairtypes[i].dtype)));
```

Comment:
Just a reminder that we need better comments, something like:
```
/* built-in pairtypes have fixed handle values, therefore their index are also fixed. To make it work with MPIR_Datatype_direct, we need ensure that index matches the actual index. 
 */
```
Actually, how about a direct assertion:
```
 MPIR_Assert(HANDLE_INDEX(d) == i);
```
?

Reply:
Added in new commit.

Reply:
Added in new commit.